### PR TITLE
Tests for Thenable-curtailment

### PR DESCRIPTION
**Pull request by mgaudet**
I just want to call out the commit message on c3386c6, esp. in light of #5115 

> These tests were authored with the assistance of an LLM, however they were 
> targeted and reviewed by myself. 
> 
> The tests were validated on the prototype implemenation of SafePromiseResolve 
> in the Firefox tree; the tests highlight a current divergence between the 
> SpiderMonkey implementation and the spec-text as written (specfically 
> SpiderMonkey will currently defer on a `then` setter even with no getter, 
> which is not correct by specification).

(Of course, seeing that in the Github editor I realize how badly I typo'd some of that message 🤦 )

### CODE REVIEW

**Comment by mgaudet** [comment]
(See #5122 for the testing plan)

File: INTERPRETING.md [hunk 1]
```diff
@@ -81,6 +81,7 @@ properties of the global scope prior to test execution.
           Use this property to test that ECMAScript algorithms aren't mis-implemented to treat `document.all` as being `undefined` or of type Undefined (instead of Object).
 
           **Tests using this function must be tagged with the `IsHTMLDDA` feature so that only hosts supporting this property will run them.**
+  - **`safeResolvePromise`** - (present only in implementations that can provide it) a function which takes as its first argument a promise and as its second argument a value, and uses the SafeResolve semantics provided by the [Thenable Curtailment](https://github.com/tc39/proposal-thenable-curtailment) proposal to resolve the provided promise with the value.
```
Comment (bakkot) [inline INTERPRETING.md:84]: This signature doesn't work.

Lots of the tests rely on the fact that a Promise's state becomes locked to follow another thenable once its original resolver has been invoked. The thing is, at least at the spec level, this locking is a property of _the resolvers that are created when the Promise was originally created_, not of the Promise itself; that's why the later resolvers that get passed to the settlement value's `.then` can settle the Promise even though the original resolvers no longer can (and those resolvers have their own lock bit in case that later resolver is passed a thenable, and so on).

So this needs to be passed the resolve function from the original capability. The most straightforward thing would be to pass it a `{ promise, resolve }` pair (probably with a note that any other properties should be ignored, so it can be passed the result of `Promise.withResolvers` directly).

File: test/built-ins/Promise/safe-resolve-promise/deferred-proxy-get-trap.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  "tick 2",
+  "settled",
+];
+
+var actual = [];
+
+// "then" is the only property read from the resolution, so any other lookup
+// shows up in the comparison below as an unexpected entry.
+var value = new Proxy({}, {
+  get: function(target, key) {
```
Comment (bakkot) [inline test/built-ins/Promise/safe-resolve-promise/deferred-proxy-get-trap.js]: We should test all the traps, not just `get`. No need to write them out manually, just something along the lines of

```js
var value = new Proxy({}, new Proxy({}, {
  get: function (_, trap) {
    return function (...args) {
      actual.push({ trap, args });
      if (trap === 'get' && args[1] === 'then') {
        return function (resolve) {
          actual.push("call then");
          resolve("from the trap");
        };
      }
      return Reflect[trap](...args);
    };
  }
}));
```
I don't think this ends up changing the expectations.

File: test/built-ins/Promise/safe-resolve-promise/deferred-proxy-on-prototype-chain.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+features: [thenable-curtailment, safeResolvePromise, promise-with-resolvers, Proxy]
+---*/
+
+var actual = [];
+
+// The Proxy is the immediate prototype, and further up the chain behind
+// ordinary objects.
+function makeValue(label, depth) {
+  var proxyProto = new Proxy({}, {
+    get: function(target, key, receiver) {
```
Comment (bakkot) [inline test/built-ins/Promise/safe-resolve-promise/deferred-proxy-on-prototype-chain.js]: Ditto above re other traps.

**Review by bakkot — COMMENTED** [review]
Looks generally good except that the hook has the wrong signature. I think it should be straightforward to fix that with an LLM though. Also a couple gaps for which I have filed https://github.com/mgaudet/test262/pull/1.

I confess I did not trace through the timing for the timing tests. I'm just going to trust you got that right.

Mapping the original test plan:

Synchronous Resolve
- [x] Primitive values resolve synchronously: [sync-non-object.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-non-object.js)
- [x] Object with no `then` property resolve synchronously: [sync-no-then.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-no-then.js)
- [x] Object with un-callable `then` resolve synchronously: [sync-then-not-callable.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-then-not-callable.js)
- [x] Object with un-callable `then` on prototype resolves synchronously: [sync-then-not-callable-proto.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-then-not-callable-proto.js)
- [x] Object with `then` setter, but no getter, resolve synchronously: [sync-then-setter-only.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-then-setter-only.js)
- [x] Object with `then` setter, but no getter, on the prototype. resolves synchronously: [sync-then-setter-only-proto.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-then-setter-only-proto.js)
- [x] Object with own-property then which is not callable shadows a callable then, resolves synchronously: [sync-own-non-callable-then-shadows-callable.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-own-non-callable-then-shadows-callable.js)

Deferred Resolve & Execution of User Code
- [x] Resolving with a `Proxy` (even with empty traps) resolves deferred: [deferred-proxy-empty-handler.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-proxy-empty-handler.js)
- [x] Resolving with a Proxy with a `get` trap is resolved: [deferred-proxy-get-trap.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-proxy-get-trap.js)
  - [x] The get trap runs in deferral: [deferred-proxy-get-trap.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-proxy-get-trap.js)
- [x] Resolving an object which has a proxy on its prototype chain is deferred: [deferred-proxy-on-prototype-chain.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-proxy-on-prototype-chain.js)
- [x] Resolving with a module namespace object is deferred: [deferred-module-namespace.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-module-namespace.js); fixture: [deferred-module-namespace_FIXTURE.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-module-namespace_FIXTURE.js)
- [x] Object with callable `then` own-property is deferred: [deferred-then-callable-own.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-callable-own.js)
  - [x] The `then` call is made, deferred: [deferred-then-callable-own.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-callable-own.js)
- [x] Object with callable `then` prototype is deferred: [deferred-then-callable-proto.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-callable-proto.js)
  - [x] The then call is made deferred: [deferred-then-callable-proto.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-callable-proto.js)
- [x] A deferred `then` throwing should cause the promise to be rejected: [deferred-then-call-throws.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-call-throws.js)
- [x] Object with `then` getter: [deferred-then-getter-own.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-getter-own.js)
  - [x] The then getter is run, deferred: [deferred-then-getter-own.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-getter-own.js)
- [x] Object with `then` getter on prototype: [deferred-then-getter-proto.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-getter-proto.js)
  - [x] The then getter is run, deferred: [deferred-then-getter-proto.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-getter-proto.js)
- [x] `then` getter which throws rejects promise, but deferred: [deferred-then-getter-throws.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-getter-throws.js)
- [x] Resolving with a revoked proxy rejects deferred: [deferred-proxy-revoked.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-proxy-revoked.js)
- [x] Callable Object.prototype.then causes object resolution to be deferred: [deferred-object-prototype-then.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-object-prototype-then.js)
  - [x] unless the object has a null prototype: [sync-null-proto-ignores-object-prototype-then.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-null-proto-ignores-object-prototype-then.js)
- [x] Ensure deferred `then` call has correct this binding: [deferred-then-this-binding.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-this-binding.js)
- [x] Tick count of safe resolving a promise with another native promise (resolved, unresolved): [deferred-native-promise-ticks.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-native-promise-ticks.js), [deferred-native-promise-pending-ticks.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-native-promise-pending-ticks.js), [deferred-native-promise-pending-through-assimilation-ticks.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-native-promise-pending-through-assimilation-ticks.js)
- [x] Delete the callable `then` after a safeResolve job is enqueued but before it's processed: [deferred-then-deleted-before-job.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-then-deleted-before-job.js)
- [ ] Ensure the `Get` of `then` is run twice when: **got cut off, no idea what this was**.

Latching behaviour
- [x] Attempting to resolve a deferred promise of any sort cannot overwrite or otherwise interfere with the deferred promise: [latch-after-deferred.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-after-deferred.js), [latch-thenable-resolving-functions.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-thenable-resolving-functions.js), [latch-thenable-resolving-functions-pending.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-thenable-resolving-functions-pending.js)
- [x] Late rejection also fails: [latch-after-deferred.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-after-deferred.js), [latch-after-sync.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-after-sync.js), [latch-thenable-resolving-functions.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-thenable-resolving-functions.js), [latch-thenable-resolving-functions-pending.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-thenable-resolving-functions-pending.js)
- [x] Repeated safe resolution: Latched by first and second is no-op: [latch-double-safe-resolve.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-double-safe-resolve.js)
- [x] Attempting to safe resolve an already resolved promise is a no-op: [latch-already-settled.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/latch-already-settled.js)

Edge Cases
- [ ] (If possible) test HostObjectHasImpureLookupInternalMethods: **not really testable**
- [x] Self-resolve: (note: Spec currently will defer depending on the inspection of current promise.. but I actually think maybe we should change the spec to synchronously throw here?): [sync-self-resolution.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/sync-self-resolution.js), [deferred-self-resolution.js](https://github.com/mgaudet/test262/blob/thenable-curtailment-testing/test/built-ins/Promise/safe-resolve-promise/deferred-self-resolution.js). These cover the current spec I think.

File: test/built-ins/Promise/safe-resolve-promise/latch-after-ordinary-pending.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  var thenCallCount = 0;
+
+  return Promise.all([
+    checkIgnoredResolution("non-thenable", function() {
+      return "must be ignored";
+    }),
+
+    checkIgnoredResolution("thenable", function() {
+      return {
+        then: function(resolve) {
```
Comment (bakkot) [inline test/built-ins/Promise/safe-resolve-promise/latch-after-ordinary-pending.js:70]:
```suggestion
        get then() {
```
Might as well assert it's not even touched, not just that it's not called.

**Review by bakkot — APPROVED** [review]
LGTM other than nit.
